mono/metadata/object.c

```diff
@@ -3955,36 +3955,46 @@ mono_runtime_get_main_args (void)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 	MonoError error;
-	MonoArray *result = mono_runtime_get_main_args_checked (&error);
+	MonoArrayHandle arg_array = mono_runtime_get_main_args_handle (&error);
+	MonoArray *result = MONO_HANDLE_RAW (arg_array);
 	mono_error_assert_ok (&error);
 	return result;
 }
 
+static MonoStringHandle
+string_from_main_arg (MonoDomain *domain, int idx, MonoError *error)
+{
+	HANDLE_FUNCTION_ENTER ();
+	error_init (error);
+	MonoStringHandle res = mono_string_new_handle (domain, main_args [idx], error);
+	HANDLE_FUNCTION_RETURN_VAL (res);
+}
+
 /**
- * mono_runtime_get_main_args_checked:
+ * mono_runtime_get_main_args_handle:
  * \param error set on error
  * \returns a \c MonoArray with the arguments passed to the main
  * program. On failure returns NULL and sets \p error.
  */
-MonoArray*
-mono_runtime_get_main_args_checked (MonoError *error)
+MonoArrayHandle
+mono_runtime_get_main_args_handle (MonoError *error)
 {
-	MonoArray *res;
+	HANDLE_FUNCTION_ENTER ();
+	MonoArrayHandle res;
 	int i;
 	MonoDomain *domain = mono_domain_get ();
-
 	error_init (error);
 
-	res = (MonoArray*)mono_array_new_checked (domain, mono_defaults.string_class, num_main_args, error);
-	return_val_if_nok (error, NULL);
+	res = mono_array_new_handle(domain, mono_defaults.string_class, num_main_args, error);
+	mono_error_assert_ok (error);
 
 	for (i = 0; i < num_main_args; ++i) {
-		MonoString *arg = mono_string_new_checked (domain, main_args [i], error);
-		return_val_if_nok (error, NULL);
-		mono_array_setref (res, i, arg);
+		MonoStringHandle arg_string = string_from_main_arg (domain, i, error);
+		mono_error_assert_ok (error);
+		MONO_HANDLE_ARRAY_SETREF (res, i, arg_string);
 	}
 
-	return res;
+	HANDLE_FUNCTION_RETURN_VAL (res);
```

Comment:
The way to think about `HANDLE_FUNCTION_ENTER`/`HANDLE_FUNCTION_RETURN_*` is that they push/pop a handle stack mark: on enter the current handle stack pointer is saved and on leave it is restored back to the saved mark.  That means any `MONO_HANDLE_NEW` calls in-between will become invalid after the RETURN.

In the code above, `res` was allocated on the handle stack of `mono_runtime_get_main_args_handle`, so when you return it with `RETURN_VAL`, you're returning a garbage pointer because `res` is no longer allocated after a return.

What you want is `HANDLE_FUNCTION_RETURN_REF` - it will allocate a new handle in the *callers* handle stack to hold the result.